src/ioctls/vcpu.rs

```diff
@@ -1496,6 +1511,143 @@ impl VcpuFd {
+    ///
+    /// # Arguments
+    ///
+    /// * `reg` - The [`SyncReg`] to copy into the guest
+    ///
+    /// # Example
+    ///
+    ///  ```rust
+    /// # extern crate kvm_ioctls;
+    /// # use kvm_ioctls::{Kvm, SyncReg};
+    /// let kvm = Kvm::new().unwrap();
+    /// let vm = kvm.create_vm().unwrap();
+    /// let mut vcpu = vm.create_vcpu(0).unwrap();
+    /// vcpu.set_sync_dirty_reg(SyncReg::Register);
+    /// ```
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    pub fn set_sync_dirty_reg(&mut self, reg: SyncReg) {
+        let mut kvm_run: &mut kvm_run = self.kvm_run_ptr.as_mut_ref();
+        dbg!(kvm_run.kvm_dirty_regs, reg);
+        kvm_run.kvm_dirty_regs |= reg as u64;
+        dbg!(kvm_run.kvm_dirty_regs);
+    }
+
+    /// Disable the given [`SyncReg`] to be copied to userspace on the next exit
+    ///
+    /// # Arguments
+    ///
+    /// * `reg` - The [`SyncReg`] to not copy out of the guest
+    ///
+    /// # Example
+    ///
+    ///  ```rust
+    /// # extern crate kvm_ioctls;
+    /// # use kvm_ioctls::{Kvm, SyncReg};
+    /// let kvm = Kvm::new().unwrap();
+    /// let vm = kvm.create_vm().unwrap();
+    /// let mut vcpu = vm.create_vcpu(0).unwrap();
+    /// vcpu.clear_sync_valid_reg(SyncReg::Register);
+    /// ```
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    pub fn clear_sync_valid_reg(&mut self, reg: SyncReg) {
+        let mut kvm_run: &mut kvm_run = self.kvm_run_ptr.as_mut_ref();
+        kvm_run.kvm_valid_regs &= !(reg as u64);
+    }
+
+    /// Tell KVM to not copy the given [`SyncReg`] into the guest on the next entry
+    ///
+    /// # Arguments
+    ///
+    /// * `reg` - The [`SyncReg`] to not copy out into the guest
+    ///
+    /// # Example
+    ///
+    ///  ```rust
+    /// # extern crate kvm_ioctls;
+    /// # use kvm_ioctls::{Kvm, SyncReg};
+    /// let kvm = Kvm::new().unwrap();
+    /// let vm = kvm.create_vm().unwrap();
+    /// let mut vcpu = vm.create_vcpu(0).unwrap();
+    /// vcpu.clear_sync_dirty_reg(SyncReg::Register);
+    /// ```
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    pub fn clear_sync_dirty_reg(&mut self, reg: SyncReg) {
+        let mut kvm_run: &mut kvm_run = self.kvm_run_ptr.as_mut_ref();
+        kvm_run.kvm_dirty_regs &= !(reg as u64);
+    }
+
+    /// Get the [`kvm_sync_regs`] from the VM
+    ///
+    /// # Example
+    ///
+    ///  ```rust
+    /// # extern crate kvm_ioctls;
+    /// # use kvm_ioctls::{Kvm, SyncReg};
+    /// let kvm = Kvm::new().unwrap();
+    /// let vm = kvm.create_vm().unwrap();
+    /// let mut vcpu = vm.create_vcpu(0).unwrap();
+    /// vcpu.set_sync_valid_reg(SyncReg::Register);
+    /// vcpu.run();
+    /// let guest_rax = vcpu.sync_regs().regs.rax;
+    /// ```
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    pub fn sync_regs(&self) -> kvm_sync_regs {
+        let kvm_run: &mut kvm_run = self.kvm_run_ptr.as_mut_ref();
+        unsafe { kvm_run.s.regs }
```

Comment:
Unsafe code needs to have a comment about why is it safe. You can take a look at some examples here:
- https://github.com/rust-vmm/kvm-ioctls/blob/97ff779b6ea96ae451308149b0ae38f13b271c19/src/ioctls/vcpu.rs#L1347
- https://github.com/rust-vmm/kvm-ioctls/blob/97ff779b6ea96ae451308149b0ae38f13b271c19/src/ioctls/vcpu.rs#L1223

The unsafeness in this case comes from accessing the union field, so we need to write about why is it safe to access.

Reply:
From the perspective of an arbitrary vcpu, I'm not sure we have the safety guarantee that the `kvm_run` pointer was `mmap`'ed with enough space for the `sync_regs` itself. 

I was looking at adding a sanity check that the `KvmRunWrapper.mmap_size` is large enough to fill the `kvm_run` structure, but wasn't quite sure if that was the best way.

Or, could we make the both `sync_regs` and `sync_regs_mut` themselves `unsafe`, since we can't guarantee that the pointer from `kvm_run_ptr` is large enough for the entire `kvm_run` structure?

Reply:
Marking them as unsafe is a good path forward if we cannot ensure that using it is safe.

Reply:
What about the `mmap_size` check didn't work? It looks like the next best thing if 'has space for `sync_regs`' can't be built in to the type. Returning an `Result` from `sync_regs_mut` might be awkward but maybe better than the unsafe.

Reply:
Ah! `KvmRunWrapper` does keep the `mmap_size` around in the structure which we can check against. For some reason, I thought `KvmRunWrapper` could be created from an arbitrary pointer. 

I'll see about adding that check in.

My thinking was because anyone can create a `KvmRunWrapper`, then calling `sync_regs` which relies on that `kvm_run_ptr` can't be trusted (or at least isn't safe in all cases). I think adding a check against the `kvm_run_ptr.mmap_size` should suffice. I don't see a problem adding  a `Result` for that either. Only question would be looking into if the `kvm_run` parameters have changed in versions or how to get the correct offset/length for any given bindings.

Reply:
Having mmap_size be correct is guaranteed by `KvmRunWrapper`. Getting that bit of unsafe code right is already critical to the safety of the system and allows other parts to rely on the type being correct if it exists.

Reply:
Thanks for that explanation. I saw that `KvmRunWrapper` is exposed by the library, enabling the user to specify whatever `mmap_size` they wished. With that though, the user can't use this smaller `KvmRunWrapper` to construct an invalid `Vcpu`, so everything looks fine.

Thank you for spelling that out for me. Much appreciated! I'll remove those `unsafe`s.